### PR TITLE
server: metadata separation: reorganize metadata types

### PR DESCRIPTION
**Pull request by abooij**
### Description

This is an incremental PR towards #5797. It refactors metadata types that will be used to incrementally modify the metadata.

~It also adds a big type class in `Hasura.Class` that will be used as a general interface for providing a metadata API. We may want to discuss this type class in some more detail, to decide whether it follows a design pattern we'd like to include in our code base, or if it needs further thought.~ This change will be made in subsequent incremental PR.

Although this PR adds types that can hold new metadata, the intention is that it does not change what metadata actually gets stored. Any new style metadata code included should not be used. However, we should check this carefully.

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No

#### Metadata

Does this PR add a new Metadata feature?
- [x] No

#### GraphQL
- [x] No new GraphQL schema is generated

#### Breaking changes

- [x] No Breaking changes

### CODE REVIEW

File: server/src-lib/Hasura/RQL/DDL/Metadata.hs [hunk 1]
```diff
@@ -72,86 +72,12 @@ runClearMetadata _ = do
   buildSchemaCacheStrict
   return successMsg
 
-applyQP1
```
Comment (rakeshkky) [inline server/src-lib/Hasura/RQL/DDL/Metadata.hs:75]: The logic this function does has been moved to `parseJSON` of the `Metadata` type. See `parseListAsMap` [function](https://github.com/hasura/graphql-engine/pull/6103/files#diff-7b8382ab20e441fa214586be491eb6f7ca904d8b20ed97894c16a339be45219aR154).

**Review by rakeshkky — APPROVED** [review]
LGTM

**Review by tirumaraiselvan — APPROVED** [review]

File: server/src-lib/Hasura/RQL/DDL/Metadata/Generator.hs [hunk 2]
```diff
@@ -48,17 +43,26 @@ genReplaceMetadata = do
...
   where
-    genFunctionsMetadata :: MetadataVersion -> Gen FunctionsMetadata
+    genFunctionsMetadata :: MetadataVersion -> Gen Functions
     genFunctionsMetadata = \case
-      MVVersion1 -> FMVersion1 <$> arbitrary
-      MVVersion2 -> FMVersion2 <$> arbitrary
+      MVVersion1 -> OM.fromList . map (\qf -> (qf, FunctionMetadata qf emptyFunctionConfig)) <$> arbitrary
+      MVVersion2 -> arbitrary
+
+instance (Arbitrary k, Eq k, Hashable k, Arbitrary v) => Arbitrary (InsOrdHashMap k v) where
```
Comment (nicuveo) [inline server/src-lib/Hasura/RQL/DDL/Metadata/Generator.hs:51]: Having those instances here seem.... arbitrary :drum: 
Jokes aside: I think we could move those to a dedicated module... maybe `Hasura.Generators` or something? But that could be done in a separate PR.

File: server/src-lib/Hasura/RQL/Types/Permission.hs [hunk 3]
```diff
@@ -36,7 +38,7 @@ instance Q.FromCol PermType where
     "delete" -> Just PTDelete
     _        -> Nothing
 
-permTypeToCode :: PermType -> Text
+permTypeToCode :: PermType -> T.Text
```
Comment (nicuveo) [inline server/src-lib/Hasura/RQL/Types/Permission.hs:41]: nit: that's not required, we have an export of `Text` in the Prelude.

**Review by nicuveo — APPROVED** [review]
LGTM, only a few nits.

File: server/src-lib/Hasura/RQL/DDL/Metadata/Generator.hs (hunk 2 above)
Comment (abooij) [reply server/src-lib/Hasura/RQL/DDL/Metadata/Generator.hs:51]: Let me at least add a docstring/comment to this module suggesting this.

File: server/src-lib/Hasura/RQL/Types/Permission.hs (hunk 3 above)
Comment (abooij) [reply server/src-lib/Hasura/RQL/Types/Permission.hs:41]: Agreed, will fix this.

**Pull request merged**